### DISCUSSION TITLE
How to use batch_size and num_samples for HGTLoader ?

### DISCUSSION THREAD

**Discussion by LaurentBerder** [pyg-team/pytorch_geometric#6707] [General]
Hi,

I'm working with a HeteroData that modelizes bookings, looking roughly like the following for each booking:
![image](https://user-images.githubusercontent.com/19341543/218975363-b5901077-7926-4314-84b5-a30d18e6ca79.png)
Shippers, Consignees and Countries may be connected to different bookings, but a container is only linked to a single booking (though as you can see, a booking may have numerous containers).

I'd like to use the `HGTLoader` class to get subgraphs containing random bookings, but with all their direct connections (all their containers, their shipper, consignee and countries).

But I don't understand how the `num_samples` and `batch_size` arguments work in the class (I'm setting different values for the different node types, getting very different results but none looking like what I want), and the documentation doesn't explain it.

I seem to understand that `batch_size` is roughly the number of nodes (excluding `input_nodes`) that we end up with. But as far as the `num_samples`, I have no idea how to implement it.

Here's an example of results with the following arguments:
```python
loader = HGTLoader(
    data=valid_full,
    batch_size=28,
    num_samples={'booking': [1]*2, 'container': [18]*2, 'shipper': [1]*2, 'consignee': [1]*2, 'country': [3]*2},
    input_nodes='booking'
)
batch = next(iter(loader))
print({k: batch[k]['type'].shape[0] for k in node_types})
```
`{'booking': 29, 'container': 28, 'shipper': 2, 'consignee': 2, 'country': 6}`
![image](https://user-images.githubusercontent.com/19341543/218979023-af7896ea-e28e-4702-a3e7-2c9c675c9f58.png)

That looks quite promising, except I've never been able to find a setting that would gie me more than a single container per booking!

How should I approach this?

**Comment by rusty1s** [comment]
`batch_size` refers to the number of seed nodes you start sampling from. `num_samples` refers to the amount of nodes to sample for each node type in each iteration. 

It's interesting that the amount of shipper, consignee and country is so low for this. Is every booking connected to one of these?

**Reply by LaurentBerder** [reply]
Thanks for coming back to me Matthias.

Yes, each booking has a single shipper, and a single consignee, but may have between 1 and 4 countries, and the number of containers is pretty much open.

So if I understand well what you say, `batch_size` will be the number of booking nodes I end up with?
As for the `num_samples`, in the dictionary I still don't understand what these mean :
- the length of the lists
- the numbers in the lists

**Reply by rusty1s** [reply]
Ah, I see. I was wrong as well :D So `HGTLoader` is a layer-wise sampling algorithm, so `'shipper': [1]*2` would mean that you sample exactly 1 shipper in the first hop, and 1 shipper in the second hop, independent of the batch size. So if you want to sample approximately a shipper per booker in each hop, you would have to write `'shipper': [batch_size]*2`.

**Reply by LaurentBerder** [reply]
> Ah, I see. I was wrong as well :D

Glad to see I'm not the only one confused with this class...

```python
loader = HGTLoader(
    data=train_full,
    batch_size=batch_size,
    num_samples={'booking': [1]*2, 'container': [batch_size * 8]*2, 'shipper': [batch_size]*2,
                              'consignee': [batch_size]*2, 'country': [batch_size*3]*2},
    input_nodes='booking'
)
batch = next(iter(loader))
print({k: batch[k]['type'].shape[0] for k in train_full.node_types})
```
`{'booking': 3, 'container': 2, 'shipper': 4, 'consignee': 4, 'country': 3}`
![image](https://user-images.githubusercontent.com/19341543/219076340-a1237780-02dd-486c-879f-fe116e17417f.png)

I'm still confused (º～O), it all looks so random.

**Reply by rusty1s** [reply]
Random because there exists isolated nodes? This is indeed weird. `HGTLoader` should never sample isolated nodes (except a node has no neighbor).

On a second look: This might explain it. In every layer, we sample nodes for all node types, and may even sample nodes although they have a budget of zero.

**Reply by LaurentBerder** [reply]
Yesh, that is one of the reason why it seems random (bookings cannot be isolated, they at least are connected to one of each other kinds of nodes).

Plus, the fact that I specify `input_nodes = 'booking'` means that potential isolated nodes should only be `booking` nodes, not other kinds.

And why do I get so few `container` nodes, while I specify the largest amount of nodes to be picked for this kind? If anything, I should get lots of `container` nodes, quite a few `country` nodes, and only a handful of `shipper` and `consignee` nodes.

**Reply by rusty1s** [reply]
I still think this is correct. Assuming you have a `batch_size=1`, you get 3 booking nodes and 2 container nodes.

Isolated nodes can be included for every node type (the disadvantage of layer-wise sampling).

**Reply by rusty1s** [reply]
Nevermind, didn‘t see the the `batch_size = 8`. Do you can share one of your data objects so that I can take a look?

**Reply by LaurentBerder** [reply]
That's a bit sensible (and not easy to sample as that's actually what I'm trying to do).
I'll see what I can do.

**Reply by LaurentBerder** [reply]
> Isolated nodes can be included for every node type (the disadvantage of layer-wise sampling).

Any idea on how I could achieve sampling of complete bookings (meaning with all their direct dependencies) with any other non-layer-wise sampling, then?

**Reply by rusty1s** [reply]
Have you looked at `NeighborLoader` already? I think this one performs way better.

**Reply by LaurentBerder** [reply]
Here's a sample of data as a [pickle file](https://drive.google.com/file/d/1k9A_9e6EnffEcuOHAF7L_M8g1_Nf3ksG/view?usp=sharing). It should be fairly easy to read.

**Reply by LaurentBerder** [reply]
> Have you looked at `NeighborLoader` already? I think this one performs way better.

Well I have looked, but not succeeded in using it. Adapting the [code found in the documentation](https://pytorch-geometric.readthedocs.io/en/latest/modules/loader.html#torch_geometric.loader.NeighborLoader), I get results I don't understand either:

```python
from torch_geometric.loader import NeighborLoader

loader = NeighborLoader(
    valid_full,
    # Sample 2 neighbors for each node and edge type for 2 iterations
    num_neighbors={key: [2] * 2 for key in valid_full.edge_types},
    # Use a batch size of 10 for sampling training nodes of type paper
    batch_size=10,
    # Using random indices of booking nodes as input
    input_nodes=('booking', np.random.randint(valid_full['booking'].x.shape[0], size=3)),
)
```
![image](https://user-images.githubusercontent.com/19341543/220560578-10708ba4-a459-4c13-be8a-956a8220e433.png)
Here, we seem to be getting less isolated nodes, but the way nodes are selected still makes no sense to me: why would I only have 3 container nodes in the whole selection?

**Reply by rusty1s** [reply]
Thanks, let me take a look ASAP.

**Reply by LaurentBerder** [reply]
Hi @rusty1s , have you had any luck with the data so far?

**Reply by rusty1s** [reply]
Sorry, I didn't have a look yet :( Will try to do it asap.

**Reply by rusty1s** [reply]
I think your edge types may be a bit wrong. In particular, you are missing `booking->container`, `shipper->booking` and `consigness->booking` edge types.

**Reply by LaurentBerder** [reply]
I think you may have misread, @rusty1s : these edges exist. They're called, respectively, `(booking, has, container)`, `(shipper, sends, booking)` and `(consignee, receives, booking)`, and they're the first ones in the list of edges.
![image](https://user-images.githubusercontent.com/19341543/223060949-6d9f21d0-75bd-4e45-95d8-45f7bdd18fc8.png)

**Reply by rusty1s** [reply]
Oh lol, you are right. I have indeed misread - too many edge types :D 

Mh, I am running this code
```python
import pickle

from torch_geometric.loader import NeighborLoader

with open('sample.pkl', 'rb') as f:
    data = pickle.load(f)

loader = NeighborLoader(
    data,
    num_neighbors={key: [2] * 2
                   for key in data.edge_types},
    batch_size=10,
    input_nodes='booking',
)

batch = next(iter(loader))

for node_type in batch.node_types:
    print(node_type, batch[node_type].num_nodes)
for edge_type in batch.edge_types:
    print(edge_type, batch[edge_type].num_edges)
```
which outputs
```
booking 36
container 10
shipper 25
consignee 23
country 11
('booking', 'has', 'container') 10
('shipper', 'sends', 'booking') 10
('consignee', 'receives', 'booking') 10
('booking', 'from', 'country') 17
('booking', 'to', 'country') 19
('shipper', 'from', 'country') 17
('consignee', 'from', 'country') 19
('container', 'rev_has', 'booking') 10
('booking', 'rev_sends', 'shipper') 10
('booking', 'rev_receives', 'consignee') 10
('country', 'rev_from', 'booking') 10
('country', 'rev_to', 'booking') 10
('country', 'rev_from', 'shipper') 10
('country', 'rev_from', 'consignee') 10
```
which looks kinda correct to me. Can you confirm?

**Reply by LaurentBerder** [reply]
Thanks for having a look.
No, it doesn't seem correct to me, because each booking has at least 1 container, and here we only have 10 container nodes for 36 booking nodes. That's pretty much the heart of the problem I have.

**Reply by rusty1s** [reply]
I still think this is correct, as you are having a container node for every booking seed node (`batch_size=10`, sampled in the first hop), while you are getting more booking nodes in in the second hop of neighbor sampling.

**Reply by LaurentBerder** [reply]
I see. But I still don't get any additional container nodes (whilst most bookings have multiple containers).

Is it because containers are only linked to a single booking, therefore no backward hops are made?

**Reply by rusty1s** [reply]
If you want to have more container nodes (like all containers that correspond to the sampled booking nodes in the second hop), you would need to increase the number of hops to sample to three:
```
booking 58
container 36
shipper 48
consignee 46
country 23
('booking', 'has', 'container') 10
('shipper', 'sends', 'booking') 36
('consignee', 'receives', 'booking') 36
('booking', 'from', 'country') 17
('booking', 'to', 'country') 19
('shipper', 'from', 'country') 17
('consignee', 'from', 'country') 19
('container', 'rev_has', 'booking') 36
('booking', 'rev_sends', 'shipper') 25
('booking', 'rev_receives', 'consignee') 25
('country', 'rev_from', 'booking') 36
('country', 'rev_to', 'booking') 36
('country', 'rev_from', 'shipper') 25
('country', 'rev_from', 'consignee') 24

**Reply by LaurentBerder** [reply]
It still only returns 10 `('booking', 'has', 'container')` edges, and the 26 other `container` nodes are unlinked to any `booking` node in the sample (as well as a bunch of `shipper`, `consignee` and `country` nodes):
![image](https://user-images.githubusercontent.com/19341543/223424593-82d54322-72e8-4070-80e1-eae3d34270a2.png)

**Reply by LaurentBerder** [reply]
Maybe what I'm looking for isn't really a sampler...

The sampling I'm after really is only on booking nodes, and once they're sampled, I want all their direct neighbors. And if some of the neighbors are directly connected to each other, I want to keep their edges.

That's roughly it.

**Reply by rusty1s** [reply]
Mh, you are right. This still doesn't make much sense to me: `NeighborLoader` should never create isolated nodes, so there is definitely something I am missing here.

Nonetheless, if you are only interested in the induced subgraph from all sampled nodes, then you can utilize the `directed=False` option in `NeighborLoader`.

**Reply by LaurentBerder** [reply]
Huh, I guess I had overlooked the fact that `NeighborLoader` had a `directed` parameter. I'd never tried it.

It does improve quite a bit, as it doesn't return any isolated nodes (though I'm not sure what the link between this parameter and node isolation is):
![image](https://user-images.githubusercontent.com/19341543/223438862-ccf58e1c-5252-45e6-8c19-9cea7513fc2d.png)
Though I still do not have many container nodes:
```python
booking 57
container 40
shipper 50
consignee 44
country 31
('booking', 'has', 'container') 40
('shipper', 'sends', 'booking') 50
('consignee', 'receives', 'booking') 47
('booking', 'from', 'country') 55
('booking', 'to', 'country') 52
('shipper', 'from', 'country') 47
('consignee', 'from', 'country') 45
('container', 'rev_has', 'booking') 40
('booking', 'rev_sends', 'shipper') 50
('booking', 'rev_receives', 'consignee') 47
('country', 'rev_from', 'booking') 55
('country', 'rev_to', 'booking') 52
('country', 'rev_from', 'shipper') 47
('country', 'rev_from', 'consignee') 45
```

**Reply by rusty1s** [reply]
So this is for 3-hops, right? I think this is to be expected. Since you will sample new bookings in the third hop, but then never sample their container nodes since the sampling finishes after three hops. If you are running PyG from master, their is a somewhat hacky way to fix this:
```python
from torch_geometric.sampler import NumNeighbors

N = 2
num_neighbors = NumNeighbors({
  ("container", "rev_has", "booking"): [N] * 4,
}, default = [N] * 3)

loader = NeighborLoader(data, num_neighbors)
```
which basically means: "Sample three hops, but sample four hops for the `container->booking` edge type.

**Reply by LaurentBerder** [reply]
> So this is for 3-hops, right?

Yes absolutely. I see what you mean, and did expect some bookings to be incomplete for that reason. But not all of them like it's the case here. The sampler never ever seems to pick more than 1 container node per booking node.

With version 2.2.0, I get `cannot import name 'NumNeighbors' from 'torch_geometric.sampler'`.
What is the `NumNeighbors` for, and how is it different from what we're already doing with the dictionnary in `NeighborLoader`? I have tried before to customize it with something along the line of:
```python
loader = NeighborLoader(
    data,
    # Sample various amounts of neighbors for various edge types for 3 hops
    num_neighbors={('booking', 'has', 'container'): [-1]*3,
                   ('shipper', 'sends', 'booking'): [2]*3,
                   ('consignee', 'receives', 'booking'): [2]*3,
                   ('booking', 'from', 'country'): [2]*3,
                   ('booking', 'to', 'country'): [2]*3,
                   ('shipper', 'from', 'country'): [2]*3,
                   ('consignee', 'from', 'country'): [2]*3,
                   ('container', 'rev_has', 'booking'): [-1]*3,
                   ('booking', 'rev_sends', 'shipper'): [2]*3,
                   ('booking', 'rev_receives', 'consignee'): [2]*3,
                   ('country', 'rev_from', 'booking'): [2]*3,
                   ('country', 'rev_to', 'booking'): [2]*3,
                   ('country', 'rev_from', 'shipper'): [2]*3,
                   ('country', 'rev_from', 'consignee'): [2]*3},
    # Use a batch size of 10 for sampling training nodes of type booking
    batch_size=10,
    input_nodes='booking',
    directed=False
)
```
With the use of `-1` for container-related edges (hoping to force all of them), I have, for the first time, gotten a booking with multiple container nodes:
![image](https://user-images.githubusercontent.com/19341543/223648258-5f37adac-80ba-48f4-8c28-70ee55ff1883.png)

I then did realize that I have somewhat overestimated the amount of booking nodes linked to multiple container nodes in the data sample I provided... 
![image](https://user-images.githubusercontent.com/19341543/223650980-954024af-43e9-4b58-b7bf-c442bc3fb233.png)

**Reply by rusty1s** [reply]
> how is it different from what we're already doing with the dictionnary in NeighborLoader?

You are right, it maps to the same, just easier to write and understand :)

**Reply by LaurentBerder** [reply]
Well, I think we can safely say that we solved it.

I'm now able to get complete subgraphs using `NeighborLoader`, just like I wanted it. 

Thanks for your help, sorry it involved such a lot of your time.

**Reply by rusty1s** [reply]
Glad that everything works as expected now.
